Default field attribute and derive default support

cosmikwolf opened the pull request:
Summary

  This PR implements default value support for bitfield structs, adding two major features:

  • #[default = ...] field attribute - Allows specifying default values for individual bitfield fields
  • #[derive(Default)] support - Automatically generates Default trait implementation that respects field defaults

  Features Added

  Field-Level Defaults with #[default(...)]

```rust
  #[bitfield]
  pub struct Config {
      enabled: bool,           // defaults to false (zero-initialized)
      #[default = true]
      auto_restart: bool,      // defaults to true
      #[default = 5]
      retry_count: B6,         // defaults to 5
      #[default =Mode::Auto]
      mode: Mode,             // defaults to Mode::Auto enum variant
  }
```

 Smart Constructor Behavior

  - new() - Applies field defaults while remaining const fn
  - new_zeroed() - Explicit zero-initialization (generated when defaults are used)
  - Both constructors work in const contexts

 Automatic Default Trait Implementation
```rust
  #[bitfield]
  #[derive(Default)]
  pub struct MyBitfield {
      #[default = 42]
      value: B8,
  }

  let bf = MyBitfield::default(); // Same as MyBitfield::new()
  assert_eq!(bf.value(), 42);
```
  Technical Implementation

 Const-Compatible Design

  - All default expressions must be const-evaluable (literals, const variables, enum variants)
  - Compile-time bit manipulation ensures zero runtime overhead
  - Both new() and new_zeroed() are const fn

  Example Usage
```rust
  use modular_bitfield::prelude::*;

  #[bitfield]
  #[derive(Default, Debug)]
  pub struct NetworkPacket {
      #[default = true ]
      enabled: bool,
      #[default = PacketType::Data]
      packet_type: PacketType,
      #[default = 1500]
      mtu: B12,
      payload_size: B20,  // defaults to 0
  }

  // Create with defaults applied
  let packet = NetworkPacket::default();
  assert!(packet.enabled());
  assert_eq!(packet.packet_type(), PacketType::Data);
  assert_eq!(packet.mtu(), 1500);
  assert_eq!(packet.payload_size(), 0);

  // Or explicitly zero-initialized
  let empty = NetworkPacket::new_zeroed();
```

this pull request closes #105

csnover reviewed (commented):
The test suite included in this PR is excessive and I would appreciate that getting culled down quite a lot. It is not necessary, for example, to test whether every type of literal that could go into a `syn::Expr` works—you’ve now stopped testing modular-bitfield and now are testing `syn`. It is not necessary to test multiple kinds of arithmetic—now you are testing rustc. It is not necessary to test whether defaults match constructors—there are already tests for constructors, so if default tests against literals pass, and constructor tests against literals pass, then default against constructor tests would already also pass. It is not necessary to test `Debug`—there is already a test for that and this code does not change that feature. And so on. Just write a basic test for each new mutually exclusive feature in the patch, remembering to check failure conditions, and then use code coverage analysis to find any branches that got missed.

The big problem with this patch is that it doesn’t correctly use the defaults of child types. Instead, it assumes zero-initialisation of all child types. It may not be possible to handle all `#[default(expr)]` without `const_trait_impl` but I am almost positive it is possible to correctly inherit defaults by assigning them to an associated const value of the `Specifier` trait. Then the constructor will use the overridden default value from a `#[default]` attribute if one exists, or else it will take the default value from `<#field_type as Specifier>::DEFAULT`. For structs that also derive `Specifier` the implementation of the constructor becomes just `Self { bytes: <Self as Specifier>::DEFAULT.to_le_bytes() }`.

I see this PR was closed due to feedback on #123; this PR does not have the problems that #123 had and I think is absolutely reasonable to fix up and merge. Even if you (OP) don‘t plan on working on this directly I would keep this open as a draft for anyone who might want to finish it, since this is a feature that others have requested and there is nothing totally unreasonable here, just a lot of things that need to be fixed. Thanks!

cosmikwolf commented:
I updated the main PR notes to reflect the #[default = ...] syntax. This was changed to match the #[bits = ...] syntax

cosmikwolf committed:
feat: add #[default = ...] field attribute and #[derive(Default)] support

cosmikwolf committed:
Fix formatting and regenerate .stderr for rustc 1.93.1

Fix cargo fmt issue in expand.rs and regenerate 9 upstream .stderr
files for stable rustc 1.93.1 on linux/amd64 to match CI environment.

cosmikwolf committed:
Address reviewer feedback: remove dead branch, redundant masks, rename variable

- Remove unreachable is_empty() branch in generate_const_default_bytes (#28)
- Remove redundant & 0xFF masks before u8 casts (#24)
- Rename bit_manipulations to defaults (#17)

cosmikwolf commented on docs/bitfield.md:
```diff
@@ -6,9 +6,13 @@ By default this generates the following API:
        Note that invalid bit patterns are supported in that getters and setters will
-       be protecting accesses.
+       be protecting accesses. This constructor is always `const fn`.
+    2. `new_zeroed()`: Generated when any field has a `#[default(...)]` attribute.
```
new_zeroed() provides the only straightforward way to get zero-initialized behavior after new() changes semantics due to #[default = ...].  Without it, a user would have to construct from_bytes([0u8; std::mem::size_of::<MyBitfield>()]), which is non-obvious and easy to get wrong. Since adding a #[default] attribute changes what new() returns, an explicit zero-init constructor seems warranted.

cosmikwolf committed:
Extract bit manipulation loop to const fn in private module

Move the per-field bit manipulation loop from inline generated code
to a shared set_bits_in_bytes() const fn. Uses pass-by-value array
with const generics to stay compatible with MSRV 1.76.

Also fix pre-existing clippy::identity_op warnings in generated code.

cosmikwolf committed:
Use Punctuated for offset accumulation in default codegen

Replace token stream concatenation with Punctuated<syn::Expr, Token![+]>
for building field offset expressions, matching the existing pattern in
expand_getters_and_setters. Add clippy::identity_op allow to generated
code to protect end users from false positives.
